WIP: flux-shell: support signal forwarding

grondo opened the pull request:
This PR is in progress, but since the current code is functional, I'll post it for early review.
Still needs some form of testing, but other open questions (some of which already discussed with @garlick)

 - The `job-kill` event name has the jobid embedded currently, e.g. `job-kill.<jobid>`, so that the shell only needs to subscribe to events for the current jobid. In talks with @garlick, this may have been a false economy. Easy to change now if necessary.
 - The `job-manager.kill` RPC only supports signaling all tasks in a job. @garlick had wondered  if it would be useful to allow signaling a subset of the tasks. This would be easy to implement (but would require a little extra work and more importantly extra tests)
 
Other TODO
 -  add signal handler to job shell to forward some signals, allow special handling of other signals(?) (main use case is testing in standalone mode)
 - improve job exception handling in job-exec by use of `job-kill` event (?)

Fixes #2191

dongahn commented:
> The job-manager.kill RPC only supports signaling all tasks in a job. @garlick had wondered if it would be useful to allow signaling a subset of the tasks. This would be easy to implement (but would require a little extra work and more importantly extra tests)

If this is easy enough to do, I would go for it. For computational steering, I remember LLNL's common cases were to act on the master rank or all of the rank processes. I can imagine a scheme in between in the future with coupled and/or ensembles approach being increasingly common.

garlick commented:
I think the event topic is fine as is.  We can always revist during scale testing if the number of subscriptions registered with the broker turns out to be a factor, which it probably is not.

In the interest of keeping our schedule, I'd suggest we open an issue on `flux job kill IDSET` with @dongahn's rationale quoted, and defer until we have time.

This looks great overall to me.

grondo commented:
Ok, I've pushed a set of incremental changes here, to be squashed later:
 
- Moved the `kill_all_tasks` function to `shell.h` as `flux_shell_killall()` for use from multiple shell modules
- Added local signal handlers to the shell for `SIGTERM` and `SIGINT` so that the shell itself can forward these signals to tasks instead of dying itself.
- Added a bunch of tests for `flux job kill` and the `job-manager.kill` RPC, etc.

The only TODO left here is to modify `job-exec` exception handling to use `SIGTERM` instead of `SIGKILL` to terminate a job after an exception.

garlick commented on src/shell/kill.c:
```diff
@@ -42,21 +42,11 @@ static void kill_cb (flux_t *h, flux_msg_handler_t *mh,
-        log_msg ("kill: invalid userid %u", userid);
-        return;
-    }
     if (flux_msg_unpack (msg, "{s:i}", "signum", &signum) < 0) {
```
Suggestion: `flux_event_unpack()`.

grondo replied:
Good suggestion. However, isn't `flux_msg_unpack() == flux_event_unpack()` if you don't pass storage for `topic`? (Here we don't need the topic, just the payload) And we already are guaranteed `FLUX_MSGTYPE_EVENT` because of `FLUX_MATCH_EVENT` used on msghandler.

garlick replied:
True, just for clarity I suppose, or the general argument that the event format could change.  Meh, not a big deal.

grondo replied:
Hm, good point. If event format might change, though, we might want to have `flux_event_t` instead of `flux_msg_t` to avoid confusion down the road..

garlick replied:
Not today, but point taken.

garlick commented on src/shell/kill.c:
```diff
@@ -42,21 +42,11 @@ static void kill_cb (flux_t *h, flux_msg_handler_t *mh,
-    }
     if (flux_msg_unpack (msg, "{s:i}", "signum", &signum) < 0) {
-        log_msg ("kill: invalid kill event msg payload");
+        const char *p;
```
I assume this was here to debug a problem, but it's perhaps overkill now?

grondo replied:
Yes, sorry that crept in!

garlick commented on src/cmd/flux-job.c:
```diff
@@ -391,6 +406,91 @@ int cmd_raise (optparse_t *p, int argc, char **argv)
+ *
+ * (ignoring SIGRT*)
+ */
+static const char *sigmap[] = {
```
Duplicate of `sys_siglist[]`?  (see psignal(3))

garlick replied:
oh duh, not!  (forehead slap)

grondo replied:
I did feel this was a bit hacky, and should at least wrap each definition in `#ifdef SIGXXX` like is done in coreutils `kill.c` -- however, I felt maybe this was ok for now and could be "fixed" later?

garlick replied:
Agreed.

garlick commented on src/shell/shell.c:
```diff
@@ -277,6 +285,9 @@ int main (int argc, char *argv[])
     if (!(shell.io = shell_io_create (shell.h, shell.info)))
         log_err_exit ("shell_io_create");
 
+    if (shell_pre_exec (&shell) < 0)
```
Were you wanting to move all pre-exec initialization to this function?  (Fine by me)

grondo replied:
I wasn't thinking of moving all pre-exec initialization to this function, but was thinking the `shell_pre_exec()` function as a start of a hook for plugins. I was thinking that the signals and kill-event handlers setup fit well into this model as kind of internal plugins. The other initialization being done in the shell somehow seemed more fundamental, so didn't lend itself to easily move into this function. However, maybe at least pmi initialization should go in there...

How about I tack this on to the end once I've had a chance to fully complete this feature (starting to get lost with the number of fixup commits...)

garlick replied:
Glad you're forward thinking about plugins!   Sure, however you like.

grondo commented:
Just squashed all the `fixup!` commits to make the PR a bit more sane. One or two commits still pending on the job-exec module...

grondo commented:
Also - based on a suggestion by @garlick, the `job-kill.<jobid>` event will be renamed to `shell-<jobid>.kill`, and the shell will subscribe to `shell-<jobid>.*` events at the top level, so the kill module (and any other plugin that wants to receive events) only needs to set up the msg handler for events of interest.

grondo commented:
Some update here:

 * move event subscription to shell initialization and rename event topic `shell-<id>.`
 * add a helper function for shell components that want to subscribe to subtopics and use this in the kill event code
 * avoid sending SIGKILL to job shells from job-exec module on job exceptions, instead send SIGTERM then SIGKILL after a configurable timeout.

grondo commented:
Fixed a spurious test failure on travis (probably a /bin/sh bash vs dash issue) and did a tiny bit of squashing (sorry if that disrupts someone)

If travis passes, later I will squash away the intermediate work that changed the kill event name to the more standard `shell-<id>.kill` topic name.

grondo commented:
I think this one is close to ready. I switched the synchronization for an unreliable test to use `flux kvs get --waitcreate` on the job namespace instead of `flux kvs watch -c1 --full --waitcreate --watch ..` and the test seems to be reliable now.

In order to do the above I added a `hex` argument to `flux job id --to/--from`.

garlick commented:
This is great.  Merging.

grondo commented:
Thanks!
